engine/src/main/scala/cromwell/engine/io/nio/NioFlow.scala

```diff
@@ -153,19 +166,27 @@ class NioFlow(parallelism: Int,
   }
 
   private def hash(hash: IoHashCommand): IO[String] = {
-    getHash(hash.file).map(_.hash)
+    // If there is no hash accessible from the file storage system,
+    // we'll read the file and generate the hash ourselves.
+    getStoredHash(hash.file).flatMap {
+      case Some(storedHash) => IO.pure(storedHash)
+      case None => generateMd5FileHashForPath(hash.file)
+    }.map(_.hash)
```

Comment:
This "read file to generate hash" functionality used to be part of `getHash`. I moved it out to here and renamed that method to `getStoredHash`. This resolves some weirdness in `readFileAndChecksum`, where for local files we could read the file to generate an md5, then read the file again to ensure it matched the md5 we just generated. (This never happened in practice because we only validate DRS and Blob files.)